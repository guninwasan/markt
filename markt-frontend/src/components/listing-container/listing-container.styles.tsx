import styled from "styled-components";
import { colors } from "../../utils";

const Container = styled.div`
  width: 200px;
  padding: 1rem;
  border: 1px solid ${colors.lightGrey};
  border-radius: 0.5rem;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  cursor: pointer;
  transition: transform 0.3s ease, box-shadow 0.3s ease;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  background-color: white;

  &:hover {
    transform: translateY(-4px);
    box-shadow: 0 6px 10px rgba(0, 0, 0, 0.15);
  }
`;

const Image = styled.img`
  width: 100%;
  height: auto;
  border-radius: 6px;
  object-fit: cover;
`;

const InfoContainer = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  margin-top: 0.5rem;
  justify-content: space-between;
  overflow: hidden;
`;

const Title = styled.h4`
  font-size: 16px;
  font-weight: bold;
  color: #333;
  margin: 0;
  padding: 0;
  flex: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  text-align: left;
`;

const Price = styled.span`
  font-size: 14px;
  font-weight: 1000;
  color: ${colors.darkerHoverPrimary};
  margin-left: 8px;
  overflow: hidden;
`;

const AdditionalInfo = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  margin-top: 12px;
  justify-content: space-between;
`;

const WishlistIcon = styled.div<{ isWishList?: boolean }>`
  color: ${({ isWishList }) => (isWishList ? colors.red : colors.grey)};
  cursor: pointer;
  font-size: 18px;
  transition: color 0.3s;

  &:hover {
    color: ${({ isWishList }) =>
      isWishList ? colors.darkGrey : colors.redHover};
  }
`;

const TextInfo = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  font-size: 12px;
  color: ${colors.textBlack};
  margin-left: 8px;
`;

const Condition = styled.span`
  font-size: 12px;
  color: ${colors.darkGrey};
`;

const Location = styled.span`
  font-size: 12px;
  color: ${colors.darkGrey};
  font-weight: bold;
`;

export {
  Container,
  Image,
  InfoContainer,
  Title,
  Price,
  AdditionalInfo,
  WishlistIcon,
  TextInfo,
  Condition,
  Location,
};
