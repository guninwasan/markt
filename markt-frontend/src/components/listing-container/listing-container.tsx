import { useNavigate } from "react-router-dom";
import { FaHeart } from "react-icons/fa";
import {
  Container,
  Image,
  InfoContainer,
  Title,
  Price,
  AdditionalInfo,
  WishlistIcon,
  TextInfo,
  Condition,
  Location,
} from "./listing-container.styles";

type ListingContainerProps = {
  id: string;
  image: string;
  title: string;
  price: string;
  condition: string;
  location: string;
};

const ListingContainer = ({
  id,
  image,
  title,
  price,
  condition,
  location,
}: ListingContainerProps) => {
  const navigate = useNavigate();

  const handleClick = () => {
    navigate(`/listing?id=${id}`);
  };

  return (
    <Container onClick={handleClick}>
      <Image src={image} alt={title} />
      <InfoContainer>
        <Title>{title}</Title>
        <Price>{price}</Price>
      </InfoContainer>
      <AdditionalInfo>
        <WishlistIcon>
          <FaHeart />
        </WishlistIcon>
        <TextInfo>
          <Condition>{condition}</Condition>
          <Location>{location}</Location>
        </TextInfo>
      </AdditionalInfo>
    </Container>
  );
};

export { ListingContainer };
