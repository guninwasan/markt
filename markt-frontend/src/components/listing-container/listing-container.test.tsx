import { render, screen, fireEvent } from "@testing-library/react";
import { useNavigate } from "react-router-dom";
import { ListingContainer } from "./listing-container";
import { MemoryRouter } from "react-router-dom";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: jest.fn(),
}));

describe("ListingContainer", () => {
  const mockNavigate = jest.fn();

  beforeEach(() => {
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
    mockNavigate.mockReset();
  });

  const renderListingContainer = () => {
    return render(
      <MemoryRouter>
        <ListingContainer {...defaultProps} />
      </MemoryRouter>
    );
  };

  const defaultProps = {
    id: "1",
    image: "test-image.jpg",
    title: "Test Title",
    price: "$100",
    condition: "New",
    location: "Test Location",
  };

  it("should render correctly with given props", () => {
    renderListingContainer();

    expect(screen.getByAltText("Test Title")).toHaveAttribute(
      "src",
      "test-image.jpg"
    );
    expect(screen.getByText("Test Title")).toBeInTheDocument();
    expect(screen.getByText("$100")).toBeInTheDocument();
    expect(screen.getByText("Test Location")).toBeInTheDocument();
    expect(screen.getByText("New")).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const { container } = renderListingContainer();
    expect(container).toMatchSnapshot();
  });

  it("should navigate to listing page on click", () => {
    renderListingContainer();

    fireEvent.click(screen.getByTestId("listingContainer"));
    expect(mockNavigate).toHaveBeenCalledWith("/listing?id=1");
  });

  it("should toggle wishlist state on wishlist icon click", () => {
    renderListingContainer();

    const wishlistIcon = screen.getByTestId("wishlist-button");
    fireEvent.click(wishlistIcon);
    expect(wishlistIcon).toHaveClass("wishlist-icon-active");

    fireEvent.click(wishlistIcon);
    expect(wishlistIcon).not.toHaveClass("wishlist-icon-active");
  });
});
