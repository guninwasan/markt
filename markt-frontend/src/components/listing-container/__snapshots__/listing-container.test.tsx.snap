// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingContainer should match snapshot 1`] = `
<div>
  <div
    class="sc-blHHSb kvQhZz"
    data-testid="listingContainer"
  >
    <img
      alt="Test Title"
      class="sc-gtLWhw gHqCbx"
      src="test-image.jpg"
    />
    <div
      class="sc-egkSDF iudIOR"
    >
      <h4
        class="sc-fAUdSK iAyLhe"
      >
        Test Title
      </h4>
      <span
        class="sc-dntaoT hCeDuz"
      >
        $100
      </span>
    </div>
    <div
      class="sc-ivxoEo gIDWeM"
    >
      <div
        class="sc-ghWlax cNFGyV"
        data-testid="wishlist-button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          />
        </svg>
      </div>
      <div
        class="sc-kLhKbu exAFAb"
      >
        <span
          class="sc-khLCKb blLMmS"
        >
          Test Location
        </span>
        <span
          class="sc-ixGGxD cLXaOK"
        >
          New
        </span>
      </div>
    </div>
  </div>
</div>
`;
